### PR TITLE
enh: validate alignment of sliding windows to fill with taosTimeTruncate and taosTimeCountInterval refactored

### PR DESCRIPTION
**Pull request by bgzhao66**

### CODE REVIEW

File: src/os/src/detail/osTime.c [hunk 1]
```diff
@@ -565,23 +565,29 @@ int64_t taosTimeSub(int64_t t, int64_t duration, char unit, int32_t precision) {
   return (int64_t)(mktime(&tm) * TSDB_TICK_PER_SECOND(precision));
 }
 
-int32_t taosTimeCountInterval(int64_t skey, int64_t ekey, int64_t interval, char unit, int32_t precision) {
+int64_t taosTimeTzOffset(int64_t intervalUnit, int32_t precision) {
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:568]: Extract the snippet of code to compute timezone offset in seconds as a standalone function.

File: src/os/src/detail/osTime.c [hunk 2]
```diff
@@ -565,23 +565,29 @@ int64_t taosTimeSub(int64_t t, int64_t duration, char unit, int32_t precision) {
...
+    #if defined(WINDOWS) && _MSC_VER >= 1900
+    // see https://docs.microsoft.com/en-us/cpp/c-runtime-library/daylight-dstbias-timezone-and-tzname?view=vs-2019
+    int64_t timezone = _timezone;
+    #endif
+    tz_offset = -1 * timezone * TSDB_TICK_PER_SECOND(precision);
+  }
+  return tz_offset;
+}
+
+int32_t taosTimeCountInterval(int64_t skey, int64_t ekey, int64_t sliding, int64_t slidingUnit, int64_t intervalUnit, int32_t precision) {
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:580]: Refactor taosTimeCountInterval to respect the timezone offset according to intervalUnit, keeping in sync with behaviours of taosTimeTruncate.

File: src/os/src/detail/osTime.c [hunk 3]
```diff
@@ -622,56 +628,22 @@ int64_t taosTimeTruncate(int64_t t, const SInterval* pInterval, int32_t precisio
...
-      }
-    } else {
-      end = INT64_MAX;
-    }
+    // To avoid the overly complicated effect of DST on size of sliding windows, the DST(daylight saving time) is
+    // better to be respected by users' input and output for display only, maintaining the status quo.
+    // In this way, the size of sliding windows keeps unchanging during each cycle of processing of requests.
+    start = t - pInterval->interval + pInterval->sliding;
+    int64_t tz_offset = taosTimeTzOffset(pInterval->intervalUnit, precision);
+    start -= (start+tz_offset)%pInterval->sliding;
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:646]: Truncate by subtracting the remainder within aligned intervals.

Comment (bgzhao66) [reply src/os/src/detail/osTime.c:646]: The effect of code keeps exactly the same as old, while the readability of code improves significantly.

File: src/query/src/qFill.c [hunk 4]
```diff
@@ -497,13 +507,16 @@ int64_t getNumOfResultsAfterFillGap(SFillInfo* pFillInfo, TSKEY ekey, int32_t ma
   int64_t numOfRes = -1;
   if (numOfRows > 0) {  // still fill gap within current data block, not generating data after the result set.
     TSKEY lastKey = tsList[pFillInfo->numOfRows - 1];
+    if(!validateAlignOfSlidingWindows(lastKey, pFillInfo->currentKey, pFillInfo)) return 0;
```
Comment (bgzhao66) [inline src/query/src/qFill.c:510]: Validate the alignment of sliding windows. If violated, print descriptive messages for trouble shooting along with assert coredump.

File: src/os/src/detail/osTime.c [hunk 5]
```diff
@@ -622,56 +628,22 @@ int64_t taosTimeTruncate(int64_t t, const SInterval* pInterval, int32_t precisio
 
     if (pInterval->slidingUnit == 'y') {
       tm.tm_mon = 0;
-      tm.tm_year = (int)(tm.tm_year / pInterval->sliding * pInterval->sliding);
+      tm.tm_year -= (int)(tm.tm_year%pInterval->sliding);
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:631]: The effect of code is the same as old. Truncate by subtracting remainder within a sliding window, using a unified approach.

File: src/os/src/detail/osTime.c [hunk 6]
```diff
@@ -596,11 +602,11 @@ int32_t taosTimeCountInterval(int64_t skey, int64_t ekey, int64_t interval, char
...
   int emon = tm.tm_year * 12 + tm.tm_mon;
 
-  if (unit == 'y') {
-    interval *= 12;
+  if (slidingUnit == 'y') {
+    sliding *= 12;
   }
 
-  return (int32_t)(emon/interval - smon/interval) + 1;
+  return (int32_t)(emon/sliding - smon/sliding) + 1;
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:609]: The effect of code is the same as old. Truncate by subtracting remainder within a sliding window, using a unified approach.

File: src/os/src/detail/osTime.c [hunk 7]
```diff
@@ -622,56 +628,22 @@ int64_t taosTimeTruncate(int64_t t, const SInterval* pInterval, int32_t precisio
 
     if (pInterval->slidingUnit == 'y') {
       tm.tm_mon = 0;
-      tm.tm_year = (int)(tm.tm_year / pInterval->sliding * pInterval->sliding);
+      tm.tm_year -= (int)(tm.tm_year%pInterval->sliding);
     } else {
       int mon = tm.tm_year * 12 + tm.tm_mon;
-      mon = (int)(mon / pInterval->sliding * pInterval->sliding);
+      mon -= (int)(mon % pInterval->sliding);
```
Comment (bgzhao66) [inline src/os/src/detail/osTime.c:634]: The effect of code is the same as old. Truncate by subtracting remainder within a sliding window, using a unified approach.

**Review by DuanKuanJun — APPROVED** [review]

**Pull request merged**